src/main/java/io/cryostat/MainModule.java

```diff
@@ -138,36 +141,101 @@ public void println(Exception e) {
     public static Gson provideGson(Logger logger) {
-        return provideGson(Set.of(), Set.of(), logger);
+        return provideGson(
+                Set.of(new JmxServiceUrlAdapter(logger), new HttpMimeTypeAdapter()),
+                Set.of(new RuleDeserializer()),
+                logger);
+    }
+
+    // testing-only when extra deserializers aren't needed
+    public static Gson provideGson(Logger logger, PluggableTypeAdapter<?>... adapters) {
+        return provideGson(Set.of(adapters), Set.of(), logger);
+    }
+
+    // testing-only when extra adapters aren't needed
+    public static Gson provideGson(Logger logger, PluggableJsonDeserializer<?>... deserializers) {
+        return provideGson(Set.of(), Set.of(deserializers), logger);
+    }
+
+    @Provides
+    @Singleton
+    @IntoSet
+    public static PluggableTypeAdapter<?> provideJmxServiceUrlAdapter(Logger logger) {
+        return new JmxServiceUrlAdapter(logger);
+    }
+
+    @Provides
+    @Singleton
+    @IntoSet
+    public static PluggableTypeAdapter<?> provideHttpMimeTypeAdapter() {
+        return new HttpMimeTypeAdapter();
+    }
+
+    @Provides
+    @Singleton
+    @IntoSet
+    public static PluggableTypeAdapter<?> provideProbeTemplateAdapter() {
+        return new ProbeTemplateTypeAdapter();
+    }
+
+    @Provides
+    @Singleton
+    @IntoSet
+    public static PluggableJsonDeserializer<?> provideRuleDeserializer() {
+        return new RuleDeserializer();
     }
 
     // public since this is useful to use directly in tests
     @Provides
     @Singleton
     public static Gson provideGson(
-            Set<PluggableTypeAdapter<?>> extraAdapters,
+            Set<PluggableTypeAdapter<?>> typeAdapters,
+            Set<PluggableJsonDeserializer<?>> deserializers,
+            Logger logger) {
+        return gsonBuilder(typeAdapters, deserializers, logger).create();
+    }
+
+    private static GsonBuilder gsonBuilder(
+            Set<PluggableTypeAdapter<?>> typeAdapters,
             Set<PluggableJsonDeserializer<?>> deserializers,
             Logger logger) {
         GsonBuilder builder =
                 new GsonBuilder()
                         .serializeNulls()
                         .disableHtmlEscaping()
-                        .registerTypeAdapter(
-                                JMXServiceURL.class, new GsonJmxServiceUrlAdapter(logger))
+                        .registerTypeAdapter(JMXServiceURL.class, new JmxServiceUrlAdapter(logger))
                         .registerTypeAdapter(HttpMimeType.class, new HttpMimeTypeAdapter())
                         .registerTypeHierarchyAdapter(Path.class, new PathTypeAdapter())
                         .registerTypeAdapter(Rule.class, new RuleDeserializer())
                         .registerTypeAdapter(ProbeTemplate.class, new ProbeTemplateTypeAdapter())
+                        .registerTypeHierarchyAdapter(Path.class, new PathTypeAdapter())
```

Comment:
There is an duplicated `PathTypeAdapter` registered here?

Reply:
Ah thanks, I remember this chunk got weird when resolving a rebase conflict. Looks like I didn't do it right :-)